packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,314 @@
+  hideLabel: AllowedTypes.breakpoint('boolean'),
+  state: AllowedTypes.oneOf<PinCodeState>(FORM_STATES),
+  disabled: AllowedTypes.boolean,
+  loading: AllowedTypes.boolean,
+  required: AllowedTypes.boolean,
+  message: AllowedTypes.string,
+  type: AllowedTypes.oneOf<PinCodeType>(PIN_CODE_TYPES),
+  value: AllowedTypes.array(AllowedTypes.string),
+  theme: AllowedTypes.oneOf<Theme>(THEMES),
+};
+
+@Component({
+  tag: 'p-pin-code',
+  shadow: true,
+})
+export class PinCode {
+  @Element() public host!: HTMLElement;
+
+  /** The label text. */
+  @Prop() public label?: string = '';
+
+  /** The description text. */
+  @Prop() public description?: string = '';
+
+  /** Name of the control. */
+  @Prop() public name?: string;
+
+  /** Number of characters of the Pin Code. */
+  @Prop() public length?: PinCodeLength = 4;
+
+  /** Show or hide label and description text. For better accessibility it is recommended to show the label. */
+  @Prop() public hideLabel?: BreakpointCustomizable<boolean> = false;
+
+  /** The validation state. */
+  @Prop() public state?: PinCodeState = 'none';
+
+  /** Disables the Pin Code. No events will be triggered while disabled state is active. */
+  @Prop() public disabled?: boolean = false;
+
+  /** Disables the Pin Code and shows a loading indicator. No events will be triggered while loading state is active. */
+  @Prop() public loading?: boolean = false;
+
+  /** Marks the Pin Code as required. */
+  @Prop() public required?: boolean = false;
+
+  /** The message styled depending on validation state. */
+  @Prop() public message?: string = '';
+
+  /** Pin Code type. */
+  @Prop() public type?: PinCodeType = 'number';
+
+  /** Sets the initial value of the Pin Code. */
+  @Prop({ mutable: true }) public value?: string[] = [];
+
+  /** Adapts the color depending on the theme. */
+  @Prop() public theme?: Theme = 'light';
+
+  /** Emitted when selected element changes. */
+  @Event({ bubbles: false }) public update: EventEmitter<PinCodeUpdateEvent>;
+
+  private form: HTMLFormElement;
+  private isWithinForm: boolean;
+  private hiddenInput: HTMLInputElement;
+  private pinCodeElements: HTMLInputElement[] = [];
+
+  public connectedCallback(): void {
+    this.form = getClosestHTMLElement(this.host, 'form');
+    this.isWithinForm = !!this.form;
+  }
+
+  public componentWillLoad(): void {
+    if (this.isWithinForm) {
+      this.hiddenInput = initHiddenInput(this.host, this.name, this.value.join(''), this.disabled, this.required);
+    }
+  }
+
+  public componentWillRender(): void {
+    // initialize array of values with empty strings / reset initial value if it does not consist of digits only
+    if (this.value.length === 0 || (this.value.join('') && !inputConsistsOfDigits(this.value.join('')))) {
+      if (this.value.join().length > 0) {
+        warnAboutTransformedInitialValue();
+      }
+      this.value = Array(this.length).fill('');
+      this.updateValue();
```

Comment:
It looks like this is emitting the update event before anything has rendered and before `validateProps` was called?